great_expectations/experimental/datasources/spark_datasource.py

```diff
@@ -0,0 +1,229 @@
+from __future__ import annotations
+
+import copy
+import logging
+import pathlib
+import re
+from typing import TYPE_CHECKING, Dict, List, Optional, Pattern, Tuple, Type, Union
+
+import pydantic
+from typing_extensions import ClassVar, Literal
+
+import great_expectations.exceptions as ge_exceptions
+from great_expectations.alias_types import PathStr
+from great_expectations.core.batch_spec import PathBatchSpec
+from great_expectations.experimental.datasources.interfaces import (
+    Batch,
+    BatchRequest,
+    BatchRequestOptions,
+    BatchSortersDefinition,
+    DataAsset,
+    Datasource,
+)
+
+if TYPE_CHECKING:
+    from great_expectations.execution_engine import ExecutionEngine
+
+LOGGER = logging.getLogger(__name__)
+
+
+class SparkDatasourceError(Exception):
+    pass
+
+
+# TODO: <Alex>ALEX_NOTE-1/27/2023: We did not add "_pandas" suffix (nor qualified class name) for CSVSparkAsset in PandasDatasource module (we might need to do that at a later date, or come up with a different, more general, solution altogether).</Alex>
+class CSVSparkAsset(DataAsset):
+    # Overridden inherited instance fields
+    type: Literal["csv_spark"] = "csv_spark"
+
+    # Spark specific attributes
+    base_directory: pathlib.Path
+    regex: Pattern
+
+    # Internal attrs
+    _unnamed_regex_param_prefix: str = pydantic.PrivateAttr(
+        default="batch_request_param_"
+    )
+
+    def _fully_specified_batch_requests_with_path(
+        self, batch_request: BatchRequest
+    ) -> List[Tuple[BatchRequest, pathlib.Path]]:
+        """Generates a list fully specified batch requests from partial specified batch request
+
+        Args:
+            batch_request: A batch request
+
+        Returns:
+            A list of pairs (batch_request, path) where 'batch_request' is a fully specified
+            batch request and 'path' is the path to the corresponding file on disk.
+            This list will be empty if no files exist on disk that correspond to the input
+            batch request.
+        """
+        option_to_group_id = self._option_name_to_regex_group_id()
+        group_id_to_option = {v: k for k, v in option_to_group_id.items()}
+        batch_requests_with_path: List[Tuple[BatchRequest, pathlib.Path]] = []
+
+        all_files: List[pathlib.Path] = list(
+            pathlib.Path(self.base_directory).iterdir()
+        )
```

Comment:
Nit - we don't need to explicitly type here (or anywhere else where the type can be implied. `mypy` dislikes these annotations unless there is ambiguity present

Reply:
@cdkini Could you please suggest a better coding pattern?  Thank you.

Reply:
@cdkini Thank you -- I will take a look (including `Pandas`).  This code is temporary anyway.  Thanks.

Reply:
@cdkini I checked and feel that the type hint is useful here (and in `Pandas`), because of what the list contains (i.e., `pathlib.Path` is the non-obvious part here).  Thanks.